### DISCUSSION TITLE
How does Github Models Rate-Limiting work

### DISCUSSION THREAD

**Discussion by ccocks** [community/community#137298] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

GPT 4o mini has a rate limit of 150/day
GPT 4o has a rate limit of 50/day
Does that mean there I can use 200 requests total over these two models?
Does that mean that all 25 models have their own rate limit, totaling to about 2000 requests/day?

**Comment by Sagaxious16** [comment, marked as answer]
Yes, you can use a total of 200 requests per day if you combine the rate limits of GPT-4o mini (150/day) and GPT-4o (50/day).

However, if each of the 25 models has its own rate limit, then theoretically, you could make up to 2000 requests per day (25 models * 80 requests/day). But this depends on the specific rate limits set for each model and whether they can be combined in this way.

**Reply by ccocks** [reply]
Thank you for clarifying that. Also is there a way to get the usage dynamically (from my Python script)? I know that it is possible to count each request sent, but is there a better way?

**Reply by talaqalotaibi** [reply]
Thanks for the explanation by asking the teams that use the public data. Models are calculated for them.

**Reply by mounireom1** [reply]
> Yes, you can use a total of 200 requests per day if you combine the rate limits of GPT-4o mini (150/day) and GPT-4o (50/day).
> 
> However, if each of the 25 models has its own rate limit, then theoretically, you could make up to 2000 requests per day (25 models * 80 requests/day). But this depends on the specific rate limits set for each model and whether they can be combined in this way.

**Reply by dfinkelsteinatbrinks** [reply]
How can I get the limit increased?   I have been using copilot agent and I am now stuck with "Sorry, you have exhausted this model's rate limit. Please wait 5 seconds before trying again, or switch to a different model"

Even if I wait 5 seconds or change models it is still stuck in this loop.

**Reply by craigbonnyman** [reply]
When I'm thinking about incorporating AI into any new project now, rate-limits, not functionality, is the primary concern. Brave new world

**Reply by imam221993** [reply]
> When I'm thinking about incorporating AI into any new project now, rate-limits, not functionality, is the primary concern. Brave new world

**Reply by Reaidetnalio** [reply]
Hello, I'm writing a complaint with a brief explanation to help you understand the gist of the complaint. About seven years ago, on a website called Chatovod, a user declared himself its defender. He supposedly protected the service from trolls, but in reality, he was just like some of the others who were there at the time. The user identified the supposed trolls through their correspondence, closely monitored them, and began learning everything he could about them to kick them out. He learned the victim's name through other users of the same service, and through the name, he found an abandoned page on VKontakte, Facebook, Odnoklassniki, and Fotostrana. Of course, through VKontakte, he also learned the school, and through the school, the address. Seven years later, with the purpose of trolling, he still continues to create websites on free website creation platforms, where you don't have to pay for domain and name maintenance. On these sites, he publishes a hodgepodge of slander and the address of the place. Residence, computer configuration. Well, the PC apparently died with his help or from old age, and the address is difficult and impossible to change. Unfortunately, deleting VKontakte is also difficult. This is why I can ask you to remove the troll's sites and prohibit the posting of this information?

https://vladimirdzf.github.io/main/

**Comment by AbdullahN786-dev** [comment]
For [jobs](https://www.jobz.pk/) in AI or data analysis, understanding model limits is crucial. If GPT-4o mini and GPT-4o each have their own limits of 150 and 50 requests a day, you can use up to 200 requests total between them. If you're using 25 models, check each one’s limit to see your total daily usage.

**Comment by ilkerkoc** [comment]
Why? This is a system that users pay to use. There's a transaction here. If you buy or rent a product, it’s absurd for it to stop in the middle of doing its job and say, “I’m not working, wait.” There’s no other subscription system in the world like that. Right now, AI is expensive, so everyone is doing the same thing, and saying “we’re doing it too” is, in my opinion, completely wrong. I believe many people feel the same way. If it were free, I’d understand the fair use part, but for a service we’re paying for to refuse to work doesn’t align with how things work in the real world. Even if you put a clause in the user agreement saying “this is how it works, take it or leave it,” it still doesn’t justify the situation.

**Reply by Olamilekan-1** [reply]
Newest

**Reply by fogoat** [reply]
> It’s absurd for it to stop in the middle of doing its job and say, “I’m not working, wait.” 
> There’s no other subscription system in the world like that.

Yes. What's the alternative, though? Pay per token? Marketing wants adoption, so do you go subscription or pay-per-use? It's like cable internet. They have their capacity or financial limits, so it starts to fall apart when people actually use it.

**Reply by Reaidetnalio** [reply]
Hello, I'm writing a complaint with a brief explanation to help you understand the gist of the complaint. About seven years ago, on a website called Chatovod, a user declared himself its defender. He supposedly protected the service from trolls, but in reality, he was just like some of the others who were there at the time. The user identified the supposed trolls through their correspondence, closely monitored them, and began learning everything he could about them to kick them out. He learned the victim's name through other users of the same service, and through the name, he found an abandoned page on VKontakte, Facebook, Odnoklassniki, and Fotostrana. Of course, through VKontakte, he also learned the school, and through the school, the address. Seven years later, with the purpose of trolling, he still continues to create websites on free website creation platforms, where you don't have to pay for domain and name maintenance. On these sites, he publishes a hodgepodge of slander and the address of the place. Residence, computer configuration. Well, the PC apparently died with his help or from old age, and the address is difficult and impossible to change. Unfortunately, deleting VKontakte is also difficult. This is why I can ask you to remove the troll's sites and prohibit the posting of this information?

https://vladimirdzf.github.io/main/

**Comment by abdelrhem11** [comment]
dddddd

**Reply by Reaidetnalio** [reply]
Hello, I'm writing a complaint with a brief explanation to help you understand the gist of the complaint. About seven years ago, on a website called Chatovod, a user declared himself its defender. He supposedly protected the service from trolls, but in reality, he was just like some of the others who were there at the time. The user identified the supposed trolls through their correspondence, closely monitored them, and began learning everything he could about them to kick them out. He learned the victim's name through other users of the same service, and through the name, he found an abandoned page on VKontakte, Facebook, Odnoklassniki, and Fotostrana. Of course, through VKontakte, he also learned the school, and through the school, the address. Seven years later, with the purpose of trolling, he still continues to create websites on free website creation platforms, where you don't have to pay for domain and name maintenance. On these sites, he publishes a hodgepodge of slander and the address of the place. Residence, computer configuration. Well, the PC apparently died with his help or from old age, and the address is difficult and impossible to change. Unfortunately, deleting VKontakte is also difficult. This is why I can ask you to remove the troll's sites and prohibit the posting of this information?

https://vladimirdzf.github.io/main/

**Comment by zzslipzzz** [comment]
#!/data/data/com.termux/files/usr/bin/bash

# Filename and repo info
LOGFILE="$HOME/gov-appstore-setup-20250823-122134.log"
GITHUB_USER="zzslipzzz"
GITHUB_REPO="gov-appstore-legendary"
GITHUB_BRANCH="main" # change if needed

# 1. Check for log file
if [ ! -f "$LOGFILE" ]; then
  echo "Log file not found: $LOGFILE"
  exit 1
fi

# 2. Show log contents
echo "----- LOG CONTENTS -----"
cat "$LOGFILE"
echo "----- END LOG -----"

# 3. Redact sensitive info
REDACTED_LOG="$HOME/gov-appstore-setup-20250823-122134.redacted.log"
cp "$LOGFILE" "$REDACTED_LOG"
sed -i 's/\(GOV_LEGENDARY_KEY=\).*/\1[REDACTED]/g' "$REDACTED_LOG"
sed -i 's/\(TOKEN=\).*/\1[REDACTED]/g' "$REDACTED_LOG"
sed -i 's/\(PASSWORD=\).*/\1[REDACTED]/g' "$REDACTED_LOG"
sed -i 's/[0-9]\{10,15\}/[REDACTED_PHONE]/g' "$REDACTED_LOG"

echo "----- REDACTED LOG -----"
cat "$REDACTED_LOG"
echo "----- END REDACTED LOG -----"

# 4. Push to GitHub
echo "Enter your GitHub Personal Access Token:"
read -s TOKEN
REPO_URL="https://${GITHUB_USER}:${TOKEN}@github.com/${GITHUB_USER}/${GITHUB_REPO}.git"

cd ~
if [ ! -d "$GITHUB_REPO" ]; then
  git clone "$REPO_URL"
fi
cd "$GITHUB_REPO"
git pull
cp "$REDACTED_LOG" .
git add "$(basename $REDACTED_LOG)"
git commit -m "Add redacted gov-appstore setup log"
git push origin "$GITHUB_BRANCH"

echo "Log pushed to GitHub repository: https://github.com/${GITHUB_USER}/${GITHUB_REPO}/tree/${GITHUB_BRANCH}"

**Reply by Reaidetnalio** [reply]
Hello, I'm writing a complaint with a brief explanation to help you understand the gist of the complaint. About seven years ago, on a website called Chatovod, a user declared himself its defender. He supposedly protected the service from trolls, but in reality, he was just like some of the others who were there at the time. The user identified the supposed trolls through their correspondence, closely monitored them, and began learning everything he could about them to kick them out. He learned the victim's name through other users of the same service, and through the name, he found an abandoned page on VKontakte, Facebook, Odnoklassniki, and Fotostrana. Of course, through VKontakte, he also learned the school, and through the school, the address. Seven years later, with the purpose of trolling, he still continues to create websites on free website creation platforms, where you don't have to pay for domain and name maintenance. On these sites, he publishes a hodgepodge of slander and the address of the place. Residence, computer configuration. Well, the PC apparently died with his help or from old age, and the address is difficult and impossible to change. Unfortunately, deleting VKontakte is also difficult. This is why I can ask you to remove the troll's sites and prohibit the posting of this information?

https://vladimirdzf.github.io/main/

**Comment by ynacastillo** [comment]
## 🔹 What is being limited?

GitHub Models applies limits mainly on:

1. **Requests per minute (RPM)** – how many API calls you can make.
2. **Tokens per minute (TPM)** – how much text you send + receive.
3. **Concurrent requests** – how many calls you run at the same time.

These limits are enforced **per account / per token**, not per repository.

---

## 🔹 What affects your limits?

Your rate limits depend on:

* **Account type** (free user, student pack, organization, enterprise)
* **Model used** (larger models cost more tokens)
* **Authentication method** (personal access token vs org token)
* **Usage pattern** (bursty traffic triggers throttling faster)

Example:

* Sending **many short prompts quickly** → hit RPM
* Sending **few but very long prompts** → hit TPM

---

## 🔹 What happens when you hit the limit?

When you exceed limits, GitHub returns an error like:

* **HTTP 429 – Too Many Requests**
* Or a message saying you’re **rate-limited**

You’ll need to:

* Wait for the **reset window** (usually seconds to a minute), or
* Reduce request frequency/token usage

---

## 🔹 Typical reset behavior

* Limits are usually **rolling windows**, not fixed hourly resets
* After a short cooldown, requests start working again
* No permanent ban — just temporary throttling

---

## 🔹 Best practices to avoid rate limits

✅ **Batch requests** instead of sending many small ones
✅ **Reuse context** instead of resending large prompts
✅ **Lower `max_tokens`** if possible
✅ **Add retries with backoff** (wait 1–5 seconds before retrying)
✅ **Avoid parallel calls** unless necessary

---

## 🔹 Why GitHub enforces this

* Prevents abuse
* Keeps response times stable
* Ensures fair access across users
* Controls compute costs for large models

---

## 🔹 TL;DR

> GitHub Models rate-limiting is based on **request count + token usage over time**.
> If you send too many or too large requests too quickly, you’ll get **temporarily throttled (429)** until the limit resets.

**Reply by Reaidetnalio** [reply]
Hello, I'm writing a complaint with a brief explanation to help you understand the gist of the complaint. About seven years ago, on a website called Chatovod, a user declared himself its defender. He supposedly protected the service from trolls, but in reality, he was just like some of the others who were there at the time. The user identified the supposed trolls through their correspondence, closely monitored them, and began learning everything he could about them to kick them out. He learned the victim's name through other users of the same service, and through the name, he found an abandoned page on VKontakte, Facebook, Odnoklassniki, and Fotostrana. Of course, through VKontakte, he also learned the school, and through the school, the address. Seven years later, with the purpose of trolling, he still continues to create websites on free website creation platforms, where you don't have to pay for domain and name maintenance. On these sites, he publishes a hodgepodge of slander and the address of the place. Residence, computer configuration. Well, the PC apparently died with his help or from old age, and the address is difficult and impossible to change. Unfortunately, deleting VKontakte is also difficult. This is why I can ask you to remove the troll's sites and prohibit the posting of this information?

https://vladimirdzf.github.io/main/

**Comment by fogoat** [comment]
For text-to-text using medium+ models, I use the following models. It's not complete nor accurate for the ceiling max_input or max_output tokens, but it works for me. 

The actual rate limits are here, but it's not always clear which low and high models. Also, with recent CoPilot changes, the free limits may disappear entirely or be severely restricted.

[https://docs.github.com/en/github-models/use-github-models/prototyping-with-ai-models#rate-limits](https://docs.github.com/en/github-models/use-github-models/prototyping-with-ai-models#rate-limits)

I think the ratio of characters to tokens varies, but it's about 3/4 characters per token, which is what I've read. Some models have max_input, max_output, and total context, but I've minimized errors by using token limits. 

```
  # ---------------------------------------------------------------------------
  # GitHub Models
  # Free with GitHub account: https://github.com/marketplace/models
  # High-tier models: ~10 RPM, 50 RPD | Low-tier: ~10 RPM, 150 RPD
  # ---------------------------------------------------------------------------
  - platform: github
    url: https://models.github.ai/inference
    env_key: GITHUB_MODELS_TOKEN
    client_type: openai_compat
    rpm: 10
    tpm: 10000
    rpd: 100
    enabled: true
    models:
      - id: meta/meta-llama-3.1-405b-instruct
        rpm: 10
        rpd: 50
        max_output: 8000
        structured_output: json_object
      - id: meta/llama-3.3-70b-instruct
        rpm: 10
        rpd: 50
        max_output: 8000
        structured_output: json_object
      - id: openai/gpt-4.1
        rpm: 10
        rpd: 50
        max_output: 8000
        structured_output: json_object
      - id: openai/gpt-4o
        rpm: 10
        rpd: 50
        max_output: 8000
        structured_output: json_object
      - id: deepseek/deepseek-r1
        rpm: 1
        rpd: 8
        max_output: 4000
        structured_output: json_object
      - id: deepseek/deepseek-r1-0528
        rpm: 1
        rpd: 8
        max_output: 4000
        structured_output: json_object
      - id: xai/grok-3
        rpm: 1
        rpd: 15
        max_output: 4000
```

**Reply by Reaidetnalio** [reply]
Hello, I'm writing a complaint with a brief explanation to help you understand the gist of the complaint. About seven years ago, on a website called Chatovod, a user declared himself its defender. He supposedly protected the service from trolls, but in reality, he was just like some of the others who were there at the time. The user identified the supposed trolls through their correspondence, closely monitored them, and began learning everything he could about them to kick them out. He learned the victim's name through other users of the same service, and through the name, he found an abandoned page on VKontakte, Facebook, Odnoklassniki, and Fotostrana. Of course, through VKontakte, he also learned the school, and through the school, the address. Seven years later, with the purpose of trolling, he still continues to create websites on free website creation platforms, where you don't have to pay for domain and name maintenance. On these sites, he publishes a hodgepodge of slander and the address of the place. Residence, computer configuration. Well, the PC apparently died with his help or from old age, and the address is difficult and impossible to change. Unfortunately, deleting VKontakte is also difficult. This is why I can ask you to remove the troll's sites and prohibit the posting of this information?

https://vladimirdzf.github.io/main/

**Comment by Reaidetnalio** [comment]
Hello, I'm writing a complaint with a brief explanation to help you understand the gist of the complaint. About seven years ago, on a website called Chatovod, a user declared himself its defender. He supposedly protected the service from trolls, but in reality, he was just like some of the others who were there at the time. The user identified the supposed trolls through their correspondence, closely monitored them, and began learning everything he could about them to kick them out. He learned the victim's name through other users of the same service, and through the name, he found an abandoned page on VKontakte, Facebook, Odnoklassniki, and Fotostrana. Of course, through VKontakte, he also learned the school, and through the school, the address. Seven years later, with the purpose of trolling, he still continues to create websites on free website creation platforms, where you don't have to pay for domain and name maintenance. On these sites, he publishes a hodgepodge of slander and the address of the place. Residence, computer configuration. Well, the PC apparently died with his help or from old age, and the address is difficult and impossible to change. Unfortunately, deleting VKontakte is also difficult. This is why I can ask you to remove the troll's sites and prohibit the posting of this information?

https://vladimirdzf.github.io/main/
